### PR TITLE
fix(canary-web-components): skip adding slotted column placeholders when the slotted column already has data

### LINKED ISSUE #2743: [ic-data-table] Wrong columns appear when slotting in multiple columns
## Summary of the bug 

Example data structure:
```ts
    type DataRow = {
        created: number
        first: string
        middle: string
        last: string
        status: string
    }
```

When slotting in a custom element for created and status, the last name column disappears and has the contents of the status column:

image

If I remove the slotting it displays correctly, see:
image

## 🪜 How to reproduce 
Tell us the steps to reproduce the problem: 
1. View page
2. See last name data missing

## 📸 Screenshots or code 
If you can, add screenshots to show the problem. If you need to show a code example, you can include a snippet or link to a repository with an example. 

## 🖥 📱 Device 
- Type: [e.g. Desktop, mobile] 
- Device: [e.g. iPhone, MacBook, ThinkPad] 
- OS version: [e.g. macOS13, iOS16, Android 13] 
- Browser version: [e.g. Chrome 115, Safari 16] 

## 🧐 Expected behaviour 
Tell us, in enough detail, what you expected to happen instead. 

## 📝 Acceptance Criteria  
If relevant, describe in full detail the different interactions and edge cases that the component or patterns needs to fulfil. 

*Given* 
*When*  
*Then* 

## Additional info 
Tell us anything else useful to help us fix or understand the problem.

### PR DESCRIPTION
**Pull request by timja**
## Summary of the changes

There were a number of issues with how columns were laid out before with slotting this corrects them

~To take out of draft I want to take a look at:~
- [x] I think this may still be broken if data is not provided in the cells - fixed

I have tried to create a unit test for the helper but jest is only configured for a11y on this project and looks to be a pain to setup, I can try if deemed needed

## Related issue
fixes #2743

### General 

- [x] Changes to docs package checked and committed.
- [x] All acceptance criteria reviewed and met. 

### Testing

- [x] Relevant unit tests and visual regression tests added.
- [x] Visual testing against Figma component specification completed. 
- [x] Playground stories in React Storybook up to date, with any prop changes and additions addressed. 
- [x] Compare performance of modified components against develop using Performance addon in React Storybook.

### Accessibility 

- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues.
- [x] Manual screen reader testing performed using NVDA and VoiceOver.
- [x] Manual keyboard testing for keyboard controls and logical focus order.
- [x] Correct roles used and ARIA attributes used correctly where required.
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure.

### Resize/zoom behaviour 

- [x] Page can be zoomed to 400% with no loss of content.
- [x] Screen magnifier used with no issues.
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.

### System modes

- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content.
- [x] Browser support tested (Chrome, Safari, Firefox and Edge). 

### Testing content extremes

- [x] Min/max content examples tested with no loss of content or overflow.
- [x] All prop combinations work without issue.
- [x] Tested for FOUC (Flash of Unstyled Content) in both SSR (Server-Side Rendering) and SSG (Static Site Generation) settings.
- [x] Controlled and uncontrolled input components tested.
- [x] Props/slots can be updated after initial render.

### CODE REVIEW

**Comment by timja** [comment]
@GCHQ-Developer-741 would you be able to take a look as the author of the previous change please? https://github.com/mi6/ic-ui-kit/pull/2401

File: packages/canary-web-components/src/utils/helpers.ts [hunk 1]
```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
 ): IcDataTableDataType => {
   const newData: IcDataTableDataType = {};
   const newIndexes = newKeys.map((key) => key.index);
-  let controlledIndex = 0; // When a new key is added to the data, need to increment the index to account for this new object value
 
-  Object.keys(dataObject).forEach((dataKey) => {
-    if (newIndexes.includes(controlledIndex)) {
```
Comment (timja) [inline packages/canary-web-components/src/utils/helpers.ts:670]: this didn't work if e.g. new indexes were 0,1 as 1 would never be called as controlledIndex would skip over it.

Also didn't work for end columns if there were two slotted columns as the object only iterated over the existing keys and it finished iterating before it got to the end column

File: packages/canary-web-components/src/utils/helpers.ts [hunk 2]
```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
...
-      newData[newKeys[newIndexes.indexOf(controlledIndex)].key] = newValue;
-      controlledIndex++;
+  const keys = Object.keys(dataObject);
+  const values = Object.values(dataObject);
+  const numberOfKeys = keys.length + newIndexes.length;
```
Comment (timja) [inline packages/canary-web-components/src/utils/helpers.ts:670]: we now iterate over the expected amount of keys total so we don't skip keys like previously

File: packages/canary-web-components/src/utils/helpers.ts [hunk 3]
```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
...
+
+  for (let i = 0, j = 0; i < numberOfKeys; i++) {
+    if (newIndexes.includes(i)) {
+      newData[newKeys[newIndexes.indexOf(i)].key] = newValue;
+    } else {
+      newData[keys[j]] = values[j];
+      j++;
```
Comment (timja) [inline packages/canary-web-components/src/utils/helpers.ts:677]: the position in the original object that we're fetching from, we can't use i as i includes the new positions

File: packages/canary-web-components/src/utils/helpers.ts [hunk 4]
```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
 ): IcDataTableDataType => {
   const newData: IcDataTableDataType = {};
   const newIndexes = newKeys.map((key) => key.index);
-  let controlledIndex = 0; // When a new key is added to the data, need to increment the index to account for this new object value
 
-  Object.keys(dataObject).forEach((dataKey) => {
-    if (newIndexes.includes(controlledIndex)) {
-      newData[newKeys[newIndexes.indexOf(controlledIndex)].key] = newValue;
-      controlledIndex++;
+  const keys = Object.keys(dataObject);
```
Comment (evenstensberg) [inline packages/canary-web-components/src/utils/helpers.ts:668]: You should use Object.entries to get the keys and values from one expression. Use destructuring to assign them to the key, value variables. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 5]
```diff
@@ -1505,7 +1505,12 @@ export class DataTable {
           ({ key }, index) =>
             isSlotUsed(this.el, `${key}-${rowIndex}`) && { key, index }
         )
-        .filter((col) => !!col);
+        .filter(
+          (col) =>
+            !!col &&
+            // skip the column if its already in the row
+            !Object.prototype.hasOwnProperty.call(row, col.key)
```
Comment (evenstensberg) [inline packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:1512]: I'm not fan of prototype calls vs methods that already exists on the Object, but it's just a nitpick

File: packages/canary-web-components/src/utils/helpers.ts [hunk 6]
```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
...
+  const values = Object.values(dataObject);
+  const numberOfKeys = keys.length + newIndexes.length;
+
+  for (let i = 0, j = 0; i < numberOfKeys; i++) {
+    if (newIndexes.includes(i)) {
+      newData[newKeys[newIndexes.indexOf(i)].key] = newValue;
```
Comment (evenstensberg) [inline packages/canary-web-components/src/utils/helpers.ts:674]: add continue inside the if statement, and remove the else block. Will make the loop neat.

**Comment by evenstensberg** [comment]
@timja commit should look like this: `fix(canary-web-components): skip adding slotted column placeholders when the slotted column already has data`

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx (hunk 5 above)
Comment (timja) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:1512]: eslint disagrees: https://eslint.org/docs/latest/rules/no-prototype-builtins#rule-details

Comment (evenstensberg) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:1512]: Okay 👍🏾

File: packages/canary-web-components/src/utils/helpers.ts (hunk 4 above)
Comment (timja) [reply packages/canary-web-components/src/utils/helpers.ts:668]: I tried that and wasn't able to get it to work, can you provide a GitHub suggestion?

**Review by GCHQ-Developer-741 — DISMISSED** [review]
Nice work. Looks good, thanks for working on this

File: packages/canary-web-components/src/utils/helpers.ts (hunk 4 above)
Comment (evenstensberg) [reply packages/canary-web-components/src/utils/helpers.ts:668]: You'll need to use map to normalize the array. But that's more work than it should be so this is ok

**Comment by timja** [comment]
@GCHQ-Developer-299 would you be able to take a look at this please?

**Comment by timja** [comment]
/ @GCHQ-Developer-741 could you re-approve please?

**Review by GCHQ-Developer-741 — APPROVED** [review]

**Review by GCHQ-Developer-112 — APPROVED** [review]

**Pull request merged**